File: leetcode/findMedianSortedArrays/findMedianSortedArrays.cpp
#include <iostream>
#include <vector>
using namespace std;


class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        vector<double> merg;
        int sz1 = nums1.size();
        int sz2 = nums2.size();
        while(sz1 || sz2){
            if(sz1 == 0){
                merg.reserve(merg.size() + nums2.size());
                merg.insert(merg.end(), nums2.begin(), nums2.end());
                sz2 = 0;
            }
            else if(sz2 == 0){
                merg.reserve(merg.size() + nums1.size());
                merg.insert(merg.end(), nums1.begin(), nums1.end());
                sz1 = 0;
            }
            else{
                if(nums1[0] < nums2[0]){
                    merg.push_back(nums1[0]);
                    nums1.erase(nums1.begin());
                    --sz1;
                }
                else{
                    merg.push_back(nums2[0]);
                    nums2.erase(nums2.begin());
                    --sz2;
                }
            }

        }
        // for(auto x: merg)
        //     cout << x << " ";
        // cout << endl;
        int n = merg.size();
        if(n%2 != 0)
            return merg[int((n-1) / 2)];
        else
            return (merg[int(n/2)] + merg[int(n/2-1)])/2;
    }
};

int main()
{
    Solution solver;
    vector<int> nums1{1, 3};
    vector<int> nums2{2};
    cout << (solver.findMedianSortedArrays(nums1, nums2)) << endl;

    nums1 = {1, 2};
    nums2 = {3, 4};
    cout << (solver.findMedianSortedArrays(nums1, nums2)) << endl;

    nums1 = {};
    nums2 = {3, 4};
    cout << (solver.findMedianSortedArrays(nums1, nums2)) << endl;

    return 0;
}

File: leetcode/findMedianSortedArrays/findMedianSortedArrays.py
# There are two sorted arrays nums1 and nums2 of size m and n respectively.

# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

# Example 1:
# nums1 = [1, 3]
# nums2 = [2]

# The median is 2.0


# Example 2:
# nums1 = [1, 2]
# nums2 = [3, 4]

# The median is (2 + 3)/2 = 2.5

class Solution:
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        merg = list()
        while(nums1 or nums2):
            if len(nums1)==0 :
                merg = merg + nums2
                nums2 = []
            elif len(nums2)==0 :
                merg = merg + nums1
                nums1 = []
            else:
                if(nums1[0] < nums2[0]):
                    merg.append(nums1[0])
                    nums1 = nums1[1:] if len(nums1) > 1 else []
                else:
                    merg.append(nums2[0])
                    nums2 = nums2[1:] if len(nums2) > 1 else []
        n = len(merg)
        if(n%2 != 0):
            return merg[int((n-1) / 2)]
        else:
            return (merg[int(n/2)] + merg[int(n/2-1)])/2

if __name__ == '__main__':
    solver = Solution()
    nums1 = [1, 3]
    nums2 = [2]
    print(solver.findMedianSortedArrays(nums1, nums2))

    nums1 = [1, 2]
    nums2 = [3, 4]
    print(solver.findMedianSortedArrays(nums1, nums2))

    nums1 = []
    nums2 = [3, 4]
    print(solver.findMedianSortedArrays(nums1, nums2))
